common/beerocks/bcl/include/bcl/beerocks_event_loop.h

```diff
@@ -0,0 +1,165 @@
+
+#include <chrono>
+#include <functional>
+
+namespace beerocks {
+
+/**
+ * @brief The event loop is a programming construct that waits for and dispatches events 
+ * in a program.
+ *
+ * It works by making a request to some "event provider" (which generally blocks the request
+ * until an event has arrived), and then it calls the relevant event handler (i.e.: dispatches
+ * the event).
+ *
+ * The event provider used by the event loop follows the file interface, which can be selected
+ * or 'polled' (the Unix system call, not actual polling). The event loop operates
+ * asynchronously with the event originator.
+ *
+ * When the event loop forms the central control flow construct of a program, as it often does,
+ * it may be termed the main loop or main event loop. This title is appropriate because such an
+ * event loop is at the highest level of control within the program.
+ */
+template <typename T> class EventLoop {
+public:
+    /**
+     * The type of the event source (e.g. file descriptor).
+     */
+    using EventType = T;
+
+    /**
+     * The type of the event loop.
+     */
+    using EventLoopType = EventLoop<T>;
+
+    /**
+     * @brief Event handler function definition.
+     *
+     * Parameters to the event handler function are:
+     * - The shared pointer to a socket resource where the event was originated at. 
+     * Event handler can perform, for example, accept, read, write and close operations on the socket.
+     * - The event loop where it was caught. Event handler can install new handlers, remove existing
+     * handlers and even ask for loop termination.
+     * 
+     * @returns True on success or false otherwise
+     */
+    using EventHandler = std::function<bool(EventType fd, EventLoopType &loop)>;
+
+    /**
+     * Set of event handler functions, one function to handle each possible event happened.
+     * Handlers are grouped into a struct to facilitate passing them as a single parameter to the install and
+     * remove handlers functions of the event loop.
+     * Event handlers are optional and if not set for a given event, that event will be silently
+     * ignored.
+     */
+    struct EventHandlers {
+        /**
+         * Hook method that is called back by the event loop to handle read events.
+         * Read events are dispatched when the socket is ready for a read operation (a read
+         * operation will not block).
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_read;
+
+        /**
+         * Hook method that is called back by the event loop to handle write events.
+         * Write events are dispatched when the socket is ready for a write operation (a write
+         * operation will not block).
+         * @param socket Socket the event was originated at.
+         * @param loop Event loop where the event was caught on.
+         */
+        EventHandler on_write;
```

Comment:
With the risk of repeating myself: I think we should only implement what we need, when we need it. For the time being, I believe read, error and disconnect handlers are sufficient. We can add the rest when we need them.

Yes, the write handler is used in the unit test, but I don't think that that's sufficient reason to have it. I'll comment in the unit test how it can be avoided.